src/components/PupilComponents/QuizEngineProvider/QuizEngineProvider.tsx

```diff
@@ -128,8 +128,12 @@ export const QuizEngineProvider = memo((props: QuizEngineProps) => {
         (answer) => answer && correctAnswers?.includes(answer),
       );
 
-      const grade = matchingAnswers.length === correctAnswers?.length ? 1 : 0;
-
+      const grade =
+        matchingAnswers.length === correctAnswers?.length
+          ? 1
+          : matchingAnswers.length > 0
+            ? 0.5
+            : 0;
```

Comment:
Yeah something like that. Though at that point it could be extracted into a little function at the bottom of the module (with early returns instead of breaks). Means your component doesn't read like a 10 page essay. Helps anyone that comes to this down the road read your intent without having to comprehend all the code. Also keeps diffs minimal when indentation changes!